test/generator/templated_generator.cpp

```diff
@@ -0,0 +1,35 @@
+#include "Halide.h"
+
+using namespace Halide;
+
+// Define a templated generator. Normally this is a bad idea, and your template
+// parameters (e.g. the type of the input) should be GeneratorParams
+// instead. Sometimes, however, it's more convenient to have the C++ type
+// available as a template parameter. Or maybe you want to template a Generator
+// on something not expressible as a GeneratorParam. Or maybe you have a
+// deficient build system and it's difficult to specify GeneratorParams in the
+// build (note that HALIDE_REGISTER_GENERATOR_ALIAS also exists for this
+// purpose).
+template<typename T1, typename T2>
+class Templated : public Generator<Templated<T1, T2>> {
+public:
+    // A major downside of templated generators is that because we use CRTP, you
+    // must manually import names of types from the base class. For Input and
+    // Output you can also just use these equivalent globally-scoped names:
+    GeneratorInput<Buffer<T1>> input{"input", 2};
+    GeneratorOutput<Buffer<T2>> output{"output", 2};
```

Comment:
I helped someone with this exact issue on StackOverflow not too long ago... https://stackoverflow.com/a/70071173/2137996

Maybe it would be helpful to document an example of this somewhere?